Circular constraint reference in typings/utils.d.ts

**Describe the bug**
We are migrating to Envelop and ran into a very odd error when trying to boot up.
```
node_modules/typescript/lib/lib.es5.d.ts:1553:11 - error TS2313: Type parameter 'P' has a circular constraint.
```

just for reference,  this line is Pick
```
/**
 * From T, pick a set of properties whose keys are in the union K
 */
type Pick<T, K extends keyof T> = {
    [P in K]: T[P];
};
```

This was all the output and we began bashing our own code to try and find the error, finally after running `tsc --build --verbose`
we got it to print the origin inside Envelop
```
node_modules/typescript/lib/lib.es5.d.ts:1553:11 - error TS2313: Type parameter 'P' has a circular constraint.
1553     [P in K]: T[P];
               ~

  node_modules/@envelop/types/utils.d.ts:13:93
    13 export declare type Spread<A extends readonly [...any]> = A extends [infer L, ...infer R] ? SpreadTwo<L, Spread<R>> : {};
                                                                                                   ~~~~~~~~~~~~~~~~~~~~~~~
    Circularity originates in type at this location.
```

for reference, it's this line with references its own type
`Spread<A extends readonly [...any]> = A extends [infer L, ...infer R] ? SpreadTwo<L, Spread<R>> : {};`

I've tried updating the line to
```
Spread<A extends readonly [...any]> = A extends [infer L, ...infer R] ? SpreadTwo<L,R> : {};
```
Which solved the issue on our codebase. Doing the same in envelop repo produces green tests and builds. 

I can make a PR with this change but wanted to check first if it should be moved into its own type or what the purpose of it is.

**To Reproduce**
To be fair, I don't know how to reproduce it in Envelop repository, I've tried with --build --verbose and it does not give the same output. My colleague and I get the same error on our different machines with our codebase.

**Expected behavior**
No typings errors

**Environment:**

- OS:
- `"@envelop/core"`: "^1.4.0",
- `typescript` tested on `4.3.4` and `4.5.4`
- `tsc-watch`: "^4.2.8" for development mode
- NodeJS: 12.18.3, 14.17.3
- 
**Additional context**
We use a monorepo, each service have its own packages/typings

--- a/.changeset/fresh-forks-visit.md
+++ b/.changeset/fresh-forks-visit.md
@@ -1,0 +1,5 @@
+---
+'@envelop/types': patch
+---
+
+Don't use `Id` TypeScript generic utility that can cause a circular constraint error

--- a/packages/types/src/utils.ts
+++ b/packages/types/src/utils.ts
@@ -2,14 +2,10 @@ export type OptionalPropertyNames<T> = { [K in keyof T]-?: {} extends { [P in K]
 
 export type SpreadProperties<L, R, K extends keyof L & keyof R> = { [P in K]: L[P] | Exclude<R[P], undefined> };
 
-export type Id<T> = T extends infer U ? { [K in keyof U]: U[K] } : never;
-
-export type SpreadTwo<L, R> = Id<
-  Pick<L, Exclude<keyof L, keyof R>> &
-    Pick<R, Exclude<keyof R, OptionalPropertyNames<R>>> &
-    Pick<R, Exclude<OptionalPropertyNames<R>, keyof L>> &
-    SpreadProperties<L, R, OptionalPropertyNames<R> & keyof L>
->;
+export type SpreadTwo<L, R> = Pick<L, Exclude<keyof L, keyof R>> &
+  Pick<R, Exclude<keyof R, OptionalPropertyNames<R>>> &
+  Pick<R, Exclude<OptionalPropertyNames<R>, keyof L>> &
+  SpreadProperties<L, R, OptionalPropertyNames<R> & keyof L>;
 
 export type Spread<A extends readonly [...any]> = A extends [infer L, ...infer R] ? SpreadTwo<L, Spread<R>> : {};
 

--- a/packages/types/src/utils.ts
+++ b/packages/types/src/utils.ts
@@ -2,6 +2,9 @@ export type OptionalPropertyNames<T> = { [K in keyof T]-?: {} extends { [P in K]
 
 export type SpreadProperties<L, R, K extends keyof L & keyof R> = { [P in K]: L[P] | Exclude<R[P], undefined> };
 
+/** @deprecated Because it can cause https://github.com/n1ru4l/envelop/issues/1120. */
+export type Id<T> = T extends infer U ? { [K in keyof U]: U[K] } : never;
+
 export type SpreadTwo<L, R> = Pick<L, Exclude<keyof L, keyof R>> &
   Pick<R, Exclude<keyof R, OptionalPropertyNames<R>>> &
   Pick<R, Exclude<OptionalPropertyNames<R>, keyof L>> &

